File: solutions/scala/queen-attack/src/main/scala/QueenAttack.scala
import scala.math.abs

object Queen {
  private def isValidPosition(x: Int, y: Int): Boolean =
    x >= 0 && x <= 7 && y >= 0 && y <= 7

  def create(x: Int, y: Int): Option[Queen] =
    if isValidPosition(x, y) then Some(Queen(x, y)) else None
}

final case class Queen(row: Int, col: Int) {
  def canAttack(other: Queen): Boolean =
    this.row == other.row || this.col == other.col ||
      abs(this.col - other.col) == abs(this.row - other.row)
}

object QueenAttack {
  def canAttack(white: Queen, black: Queen): Boolean = white.canAttack(black)
}

File: solutions/fsharp/queen-attack/QueenAttack.fs
module QueenAttack

let create (position: int * int) =
    fst position >= 0 && fst position < 8 && snd position >= 0 && snd position < 8

let canAttack (queen1: int * int) (queen2: int * int) =
    fst queen1 = fst queen2
    || snd queen1 = snd queen2
    || abs(fst queen1 - fst queen2) = abs(snd queen1 - snd queen2)
